src/provider/provider_devdax_memory.c

```diff
@@ -0,0 +1,673 @@
+        if (tail & (page_size - 1)) {
+            tail = (tail + page_size) & ~(page_size - 1);
+        }
+
+        assert_is_page_aligned(tail, page_size);
+        assert(tail >= aligned_addr + length);
+
+        *out_addr = (void *)aligned_addr;
+        return 0;
+    }
+
+    *out_addr = ptr;
+    return 0;
+}
+
+static umf_result_t devdax_alloc(void *provider, size_t size, size_t alignment,
+                                 void **resultPtr) {
+    int ret;
+
+    if (provider == NULL || resultPtr == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    devdax_memory_provider_t *devdax_provider =
+        (devdax_memory_provider_t *)provider;
+
+    size_t page_size;
+    umf_result_t result = devdax_get_min_page_size(provider, NULL, &page_size);
+    if (result != UMF_RESULT_SUCCESS) {
+        return result;
+    }
+
+    if (alignment && (alignment % page_size) && (page_size % alignment)) {
+        LOG_ERR("wrong alignment: %zu (not a multiple or a divider of the "
+                "minimum page size (%zu))",
+                alignment, page_size);
+
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    size_t old_offset; // needed for critnib_insert()
+
+    void *addr = NULL;
+    errno = 0;
+    ret =
+        devdax_alloc_aligned(size, alignment, page_size, devdax_provider->base,
+                             devdax_provider->size, &devdax_provider->lock,
+                             &addr, &devdax_provider->offset, &old_offset);
+    if (ret) {
+        os_store_last_native_error(UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED, 0);
+        LOG_ERR("memory allocation failed");
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    // verify the alignment
+    if ((alignment > 0) && ((uintptr_t)addr % alignment)) {
+        os_store_last_native_error(UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED,
+                                   0);
+        LOG_ERR("allocated address 0x%llx is not aligned to %zu (0x%zx) "
+                "bytes",
+                (unsigned long long)addr, alignment, alignment);
+        goto err_revert;
+    }
+
+    // store (offset + 1) to be able to store offset == 0
+    ret = critnib_insert(devdax_provider->fd_offset_map, (uintptr_t)addr,
+                         (void *)(uintptr_t)(old_offset + 1), 0 /* update */);
+    if (ret) {
+        LOG_ERR("devdax_alloc(): inserting a value to the file descriptor "
+                "offset map failed (addr=%p, offset=%zu)",
+                addr, old_offset);
+    }
+
+    *resultPtr = addr;
+
+    return UMF_RESULT_SUCCESS;
+
+err_revert:
+    if (util_mutex_lock(&devdax_provider->lock)) {
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    devdax_provider->offset = old_offset;
+    util_mutex_unlock(&devdax_provider->lock);
+
+    return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+}
+
+// free() is not supported
+static umf_result_t devdax_free(void *provider, void *ptr, size_t size) {
```

Comment:
What about free function. Do we have use case for provider without free? 
Without it do this provider have any value/usecase or it is the only helper function to mmap devdax? Can we create pool from this provider?

Reply:
Yes, this provider can be used as the upstream provider of the Coarse Provider or together with jemalloc (`extent_dalloc()` always fails) that take over managing of the memory.

The design of this provider has been agreed with @pbalcer and @bratpiorka.

Reply:
I added an info about that to `README.md`.

Reply:
I still do not understand why we are doing it in this way. We can implement this provider with free, in the not too complicated way. We need bitmap with bit per page. Then we do anonymous mmap, and then map_fixed to map those pages.

off course the performance will be not great. But if we will need to make it faster we can replace this bitmap with better data structure - but it can be done latter. 
Also it should have similar performance to your version without free, if user do not use it, as we do not have fragmentation, and it is simple to find next free pages.

Reply:
I agree with @ldorau - we don't need a free in devdax provider. The upstream provider will handle the free calls from the pool.

Reply:
@lplewa you can discuss it offline also with @pbalcer